src/testdrive/src/main.rs

```diff
@@ -82,38 +82,35 @@ async fn main() {
 }
 
 async fn run(args: Args) -> Result<(), Error> {
-    let (aws_region, aws_account, aws_credentials) =
-        match (args.aws_region.parse(), args.aws_endpoint) {
-            (Ok(region), None) => {
-                // Standard AWS region without a custom endpoint. Try to find actual AWS
-                // credentials.
-                let timeout = Duration::from_secs(5);
-                let account = aws::account(timeout)
-                    .await
-                    .err_ctx("getting AWS account details")?;
-                let credentials = aws::credentials(timeout)
-                    .await
-                    .err_ctx("getting AWS account credentials")?;
-                (region, account, credentials)
-            }
-            (_, aws_endpoint) => {
-                // Either a non-standard AWS region, a custom endpoint, or both. Assume
-                // dummy authentication, and just use default dummy credentials in the
-                // default config.
-                let region = rusoto_core::Region::Custom {
-                    name: args.aws_region,
-                    endpoint: aws_endpoint.unwrap_or_else(|| "http://localhost:4566".into()),
-                };
-                let account = "000000000000";
-                let credentials = AwsCredentials::new(
-                    "dummy-access-key-id",
-                    "dummy-secret-access-key",
-                    None,
-                    None,
-                );
-                (region, account.into(), credentials)
-            }
+    let (aws_region, aws_account, aws_credentials) = if let Ok(region) = args.aws_region.parse() {
+        let region: rusoto_core::Region = region;
+        // Standard AWS region without a custom endpoint. Try to find actual AWS
+        // credentials.
+        let timeout = Duration::from_secs(5);
+        let account = aws::account(timeout)
+            .await
+            .err_ctx("getting AWS account details")?;
+        let credentials = aws::credentials(timeout)
+            .await
+            .err_ctx("getting AWS account credentials")?;
+        (region, account, credentials)
+    } else {
+        // Either a non-standard AWS region, a custom endpoint, or both. Assume
+        // dummy authentication, and just use default dummy credentials in the
+        // default config.
+        let region = rusoto_core::Region::Custom {
+            name: args.aws_region,
+            endpoint: args
+                .aws_endpoint
+                // normalize endpoint, considering an empty string to be None
+                .and_then(|e| if e == "" { None } else { Some(e) })
+                .unwrap_or_else(|| "http://localhost:4566".into()),
         };
+        let account = "000000000000";
+        let credentials =
+            AwsCredentials::new("dummy-access-key-id", "dummy-secret-access-key", None, None);
+        (region, account.into(), credentials)
+    };
```

Comment:
I'm confused about why this changed. What was wrong with the old logic?

Reply:
The `None` is overprecise: we actually want to use the endpoint specifically when the region is not valid, whether or not the endpoint is set.

In particular, that allows setting the endpoint in tests and leaving it present.

Reply:
So the idea with the old logic was that if you didn't specify an endpoint, then we'd see if the region was valid. If _either_ the region was invalid, or you specified an endpoint, then we'd take your `(region, endpoint)` pair literally.

The new code looks like it'll silently ignore the endpoint if you specify a valid AWS region, which doesn't seem right?